### ISSUE TITLE
Node v16 Released  🎉

### ISSUE THREAD

**Issue by shellscape** [aws/aws-lambda-base-images#14]
:tada: :tada: :tada: :tada:  https://nodejs.medium.com/node-js-16-available-now-7f5099a97e70

PLEASE don't wait a year to implement this in a new base image!

**Comment by zbauman3** [comment]
Would love this. Any update?

**Comment by nickreynke** [comment]
It's been nearly4 months since the v 16 release. Will AWS Lambda support v 16 in the near future? @carlzogh @krk

**Comment by say8425** [comment]
v14 dose not plan to support Apple M1.
So work in v14 is little annoying.
We need v16 ASAP.
https://github.com/nodejs/build/issues/2474#issuecomment-839683917

**Comment by PatElder** [comment]
Seems AWS only does LTS builds so according to the schedule https://nodejs.org/en/about/releases/ it'll be around November.

**Comment by Sparticuz** [comment]
> Seems AWS only does LTS builds so according to the schedule https://nodejs.org/en/about/releases/ it'll be around November.

Hopefully we'll also get [proper module support](https://stackoverflow.com/a/66688901) also.

**Comment by paulcanning** [comment]
Is there a way to create a custom runtime to use the latest version of node? If so, can anyone point me in the right direction?

**Comment by shellscape** [comment]
@paulcanning you have to use a docker image function, and the cold start on those is pretty bad.

**Comment by chengjiaapraamcos** [comment]
any warriors? :)

**Comment by Robert-Rendell** [comment]
bump

**Comment by Robert-Rendell** [comment]
Also here's the current runtimes for lambda for reference:
https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html

**Comment by shellscape** [comment]
@Robert-Rendell looks like you might be new to Github. there's a great "Edit" function for replies that wont spam subscribers to this thread. Three consecutive replies is a bit silly, please don't be that person.

**Comment by Robert-Rendell** [comment]
> @Robert-Rendell looks like you might be new to Github. there's a great "Edit" function for replies that wont spam subscribers to this issue. Three consecutive replies is a bit silly, please don't be that person.

Yes I'm an escapee from bitbucket, thank you for the tip!

**Comment by frank-unovica** [comment]
+1

**Comment by o5** [comment]
I think they are waiting for re:Invent where they will present this as big feature :)

**Comment by theo-staizen** [comment]
Github Actions as of today will start moving to Node 16 as the default for actions not specifying a version. Meanwhile AWS doesn't even provide a roadmap or any signals about this... Way to go from the market leader 🙄 

For anyone stuck waiting, to avoid any issues, you should specify a node version, when setting up node in your .yml file.

```
- uses: actions/setup-node@v2
  with:
    node-version: '14'
```

**Comment by jthomerson** [comment]
AWS: Node.js 14 is in maintenance-mode. Users deserve to know when you're going to support the long-term support version of their runtime will be available.

**Comment by paul-uz** [comment]
@o5 do you happen to know if they mentioned anything at Re:Invent?

**Comment by jessecollier** [comment]
If node 14 support timeline is any indicator, we wont see v16 til Jan-Feb 
https://aws.amazon.com/blogs/compute/node-js-14-x-runtime-now-available-in-aws-lambda/

**Comment by paul-uz** [comment]
@jessecollier hopefully, although having to wait the best part of a year to get an LTS version of Node is a literal joke.

**Comment by johan13** [comment]
> If node 14 support timeline is any indicator, we wont see v16 til Jan-Feb https://aws.amazon.com/blogs/compute/node-js-14-x-runtime-now-available-in-aws-lambda/

If [node 12 support timeline](https://aws.amazon.com/about-aws/whats-new/2019/11/aws-lambda-supports-node-js-12/) as an indicator then we should have v16 support by saturday this week. :crossed_fingers:

**Comment by ParallelUniv3rse** [comment]
@carlzogh Any updates on this please? Probably the worst thing is leaving people in the dark about this, if we know it's in 3 months, at least we can plan accordingly in our organizations.

**Comment by OG84** [comment]
+ 1

**Comment by earlybail** [comment]
wtf is the deal, this needs to have been addressed months ago

**Comment by vkmita** [comment]
The difference between the urgency that orgs like GitHub (and many other services) and AWS have with staying up to date with latest versions... AWS deserves to die

**Comment by fitzgerald1337** [comment]
if anyone on AWS team is happening to read this, Google is beating you

https://cloud.google.com/functions/docs/concepts/nodejs-runtime

![google-cloud-node-v16-support](https://user-images.githubusercontent.com/7675798/148143798-f92a9ecc-1877-49e3-8b5b-a9dd3323d3f7.jpg)

**Comment by paul-uz** [comment]
Well that's embarrassing ☝️

**Comment by revmischa** [comment]
![mr leahey disapproves](https://user-images.githubusercontent.com/245131/148213321-7bcee843-8a38-423f-8eeb-003ffaebedc7.png)

**Comment by mattghali** [comment]
sigh.. anything worth developing would be faster and easier using Tornado instead

**Comment by arunasbend** [comment]
> sigh.. anything worth developing would be faster and easier using Tornado instead

Can you guys stop spamming the thread? Tons of useless emails sent for no reason ffs

**Comment by riderx** [comment]
If you can like her comment on twitter maybe it will help us : https://twitter.com/sarutule/status/1479018085789179905?s=21

**Comment by riderx** [comment]
I had to remove my last comment, the link to the tweet I sent was link to a person not in charge of Lamba but in AWS
sorry for the false hope.

She received also few unkind message.

I agree it’s annoying we all wait for this, but please AWS, or any people working there do not deserve base messages.

the only way to make things improve is by trying to understand each other’s :)

**Comment by Sparticuz** [comment]
Not node 16.... But modules and top level await now supported in 14

https://github.com/aws/aws-cdk/issues/13274#issuecomment-1007262232

**Comment by jakub-g** [comment]
Just for context:
https://github.blog/changelog/2021-12-10-github-actions-github-hosted-runners-now-run-node-js-16-by-default/

**Comment by ahawkins** [comment]
This is a blocker for our organization. We've migrated from Node 14 to 16 for M1 support. Our lambda projects are blocked without a Node 16 runtime.

**Comment by shellscape** [comment]
@ahawkins you'll have to use docker and upload an image for the lamba functions. fwiw this will drastically increase cold start times.

**Comment by ahawkins** [comment]
@shellscape Thanks for suggesting this workaround.

**Comment by boris-cp** [comment]
> This is a blocker for our organization. We've migrated from Node 14 to 16 for M1 support. Our lambda projects are blocked without a Node 16 runtime.

We are in the same situation

**Comment by mwilson71** [comment]
It's really tough to understand why this is taking so long.

**Comment by shellscape** [comment]
@mwilson71 bureaucracy mostly. but then think about the number of services that Amazon offers, and how much of that probably tangentially runs off of lambda. then consider all of the internal business processes and serverless functions that Amazon must use to operate the retail business. combine all of that and it makes sense why something like this would take over 6 months. not saying it's right, or couldn't be prevented, but it does make some assemblance of sense.

**Comment by jthomerson** [comment]
@shellscape I don't see how that could be true ... AWS doesn't need to change from Node X to Node 16 where they might be using Node X. This is just about making an _additional_ runtime environment available. Shouldn't impact their existing services. I think we're just seeing that they don't have enough resources dedicated to this, and it's not as important to them than other things, which is more disheartening.

**Comment by shellscape** [comment]
@jthomerson if I were running an org that big, I'd want to test a new version image against _everything_. Just because it's inconvenient to us all doesn't mean there's not a reason. If you're vendor locked, be patient. No amount of hand ringing or issue comments is going to make this move any faster.

**Comment by LukeSamkharadze** [comment]
This is such a joke

**Comment by fitzgerald1337** [comment]
> @jthomerson if I were running an org that big, I'd want to test a new version image against _everything_. Just because it's inconvenient to us all doesn't mean there's not a reason. If you're vendor locked, be patient. No amount of hand ringing or issue comments is going to make this move any faster.

are you in fact Jeff Bezos in disguise doing damage control

**Comment by ofhouse** [comment]
If I had to guess, we have to wait until [Amazon Linux 2022](https://aws.amazon.com/linux/amazon-linux-2022/) becomes GA, before new runtimes for Lambda will be released.

**Comment by falnyr** [comment]
@krk @carlzogh 

Sorry, if I'm pinging the wrong people, but you are the only contributors apart from @amazon-auto. 

Could we get some clarity from AWS as in when to expect this? I am sure you understand we all need to make decisions when designing solutions for our customers. It would be great if we could get some information, please 🙏.

**Comment by Negan1911** [comment]
> PLEASE don't wait a year to implement this in a new base image!

And they did!!!

**Comment by jernejc** [comment]
An estimate would be nice at least, this is very frustrating..

**Comment by TRCSamurai** [comment]
C'mon Amazon we need a response! It's been a year!

**Comment by decentralgabe** [comment]
This is a mission critical failure. Any response from the @aws-lambda-runtimes team?

**Comment by floric** [comment]
This issue is by far the most frustrating thing about AWS so far... :/

**Comment by o5** [comment]
@Negan1911 @jernejc @TRCSamurai @decentralgabe 

You're probably new at GitHub, so please look at my screenshot. There is a function "notification subscribe" and every stupid comment like yours sends an email notification to people who are already waiting for something new.

![Screenshot 2022-01-28 at 20 22 09](https://user-images.githubusercontent.com/218562/151609502-680e4e05-cabd-495a-9682-33b591708370.png)

**Comment by kid-icarus** [comment]
> and every stupid comment like yours

stay mad @o5

**Comment by LukeSamkharadze** [comment]
@o5 LIGHT MODE SCREENSHOT!!! MY EYESS!!! You have to watchout, there are developers here

**Comment by GabrielBB** [comment]
@o5 Jesus christ, your screenshot burned my eyes. I'm using dark mode

**Comment by jtuliani** [comment]
Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
 
We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
 
If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

**Comment by chengjiaapraamcos** [comment]
> Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
> 
> We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
> 
> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

Would the v3 SDK been integrated into the image? Thanks

**Comment by jtuliani** [comment]
> Would the v3 SDK been integrated into the image? Thanks

That's an open question we're working through right now. If you have strong opinions, we'd be interested to hear them! For example, is it useful to ship the SDKv2 and SDKv3 side-by-side, so you can upgrade functions to Node 16 without switching SDK? Or is it OK to ship the SDKv3 only, and require an upgrade to Node 16 to also mean upgrading to SDKv3?  (Obviously this only applies to functions using the bundled SDK.)

**Comment by o-alexandrov** [comment]
@jtuliani It’d be better if no SDKs (v2 and v3) are included in v16+ as they increase cold starts.

However, to have a more user-friendly approach, please consider to:
- let a user to choose an image without an SDK
- and by default offer an image with SDK v3 or v2

**Comment by revmischa** [comment]
My preference would be to include the v3 SDK as long as it doesn't noticeably affect cold starts.

**Comment by alfaproject** [comment]
I'd vote to include both versions side by side, or at the very least just v3

**Comment by chengjiaapraamcos** [comment]
> > Would the v3 SDK been integrated into the image? Thanks
> 
> That's an open question we're working through right now. If you have strong opinions, we'd be interested to hear them! For example, is it useful to ship the SDKv2 and SDKv3 side-by-side, so you can upgrade functions to Node 16 without switching SDK? Or is it OK to ship the SDKv3 only, and require an upgrade to Node 16 to also mean upgrading to SDKv3? (Obviously this only applies to functions using the bundled SDK.)

It depends on the affect on cold start. If not much then no brainer - both or V3.. otherwise better to be none of them.

**Comment by jamspaceteam** [comment]
I think it's best not to have surprises: I'm expecting an image without pre-existing packages. Plus, this means that it's easier to increment the version in case the SDK has security vulnerabilities

**Comment by flavianh** [comment]
I think it's best not to have surprises: I'm expecting an image without pre-existing packages. Plus, this means that it's easier to increment the version in case the SDK has security vulnerabilities

**Comment by paul-uz** [comment]
> Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
> 
> We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
> 
> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

@jtuliani thank you for an official update, but could we please get an actual answer as to why this is taking so long? And will we have to wait as long when the next node version hits LTS for another new image?

**Comment by flavianh** [comment]
@jtuliani thanks for the apology. I feel the community has been kept in the dark for a while (9 months) regarding the release of this image. This led to confusion in the community that only increased when NodeJS 14 went into maintenance mode.
Your comment underlines that closing the thread was a mistake and that you're going to find out why, and that is already great. 
I'm more interested into the bigger issue though. What's your analysis of the (high) delay between Node 16's release and the  availability of a new base image?

**Comment by lzaugg** [comment]
> I think it's best not to have surprises: I'm expecting an image without pre-existing packages. Plus, this means that it's easier to increment the version in case the SDK has security vulnerabilities

Fully agree. Additionally, users who are using ES modules and a bundler with tree-shaking at build time already, do not need (or even do not want) the SDK in the base image imo. Maybe a solution to tackle the different worlds could be to provide different lambda layers for the SDKs. Just my2c.

**Comment by eduardoboucas** [comment]
I would also be in favour of not including the SDK in the image, allowing folks to bundle the version they want to use and keeping the image lean for those who don't need the SDK.

**Comment by EHadoux** [comment]
I would advocate for having v2 in Node <= 14 and v3 in Node >= 16. I assume that the sdk is contained in a layer so there seems to be no impact on the cold start by having it (see e.g. https://medium.com/consulner/performance-of-aws-lambda-with-and-without-layers-9bffbb5434f3). However, not having the sdk in the image means that we have to bundle it with the code, which means a bigger package, which in turn means bigger cold start time. Am I wrong on that?

**Comment by o-alexandrov** [comment]
[This article by Yan Cui](https://medium.com/lumigo/this-is-all-you-need-to-know-about-lambda-cold-starts-a4982fe332e0) should help to understand factual implications of an SDK and bundling only necessary parts of it on cold starts.

The reason why the layer didn’t affect cold starts noticeably [in the comment above](https://github.com/aws/aws-lambda-base-images/issues/14#issuecomment-1030012008) is (quoting the author):
> In my case the layer contained only some static content (jar files in particular)

- so in the tests of the layer above the function doesn’t actually require/import anything from the layer as unused static files were included, thus not affecting the cold starts

**Comment by EHadoux** [comment]
> so in the tests of the layer above the function doesn’t actually require/import anything from the layer as unused static files were included, thus not affecting the cold starts

So with this line of reasoning, the SDK will act as unused static files for people not using it (because it won't be imported), thus not affecting the cold start for them.

For the others, only what is used will be imported in both cases (bundled or provided), but with a provided SDK the package will be smaller, reducing the deployment time.

**Comment by o-alexandrov** [comment]
@EHadoux please consider to familiarize with tree-shaking and therefore the fact that you cannot do tree-shaking on the provided SDK, thus importing more than needed affecting actually both cold starts and performance.

With sdk v2 the problem is even bigger than with v3 as it’s modularity is worse.

You should forget about comparisons with static files and consider cases when an SDK is used.

But coming back to even static files, it’s theoretically impossible to claim that bigger (size-wise) containers would load exactly the same as smaller ones.

**Comment by JordanMcDonough** [comment]
I came here looking for a way to run a lambda function in docker with a node 16 base image. I'm not entirely sure if that's what these images are for of if this is the wrong place to post this feel free to remove this comment.

Otherwise if anyone needs a solution in the meantime until an official base image is create, this is works for me:

```
# Define custom function directory
ARG FUNCTION_DIR="/function"

########################### Base ###########################
FROM node:16.13.2-buster-slim as base
ARG FUNCTION_DIR

# Make sure the directory exists and set it as the default for all future images
RUN mkdir -p ${FUNCTION_DIR}
WORKDIR ${FUNCTION_DIR}

########################### Lambda runtime ##############################
# This step is required for the docker container to be executed on lambda.
# We do this as a seperate build stage to reduce the size of the output
# docker image and also to seperate the logic specific to lambda
FROM base as lambda-installer

# Install aws-lambda-cpp build dependencies
RUN apt-get update && \
    apt-get install -y \
            autoconf \
            automake \
            cmake \
            g++ \
            libcurl4-openssl-dev \
            libtool \
            make \
            python3 \
            unzip

# Builds and installs aws-lambda-ric
RUN npm install aws-lambda-ric

# Uncomment the below if you want to install sharp
# RUN npm config set ignore-scripts true
# RUN npm install --platform=linux --arch=x64 sharp --unsafe-perm

########################### Installer ##############################
FROM base as installer

# Copies the files needed for installing
COPY ./package.json ./

# Install
RUN npm install --production

########################### Final Lambda App ###########################
FROM base

# Copy in the lambda node modules
COPY --from=lambda-installer ${FUNCTION_DIR}/node_modules ${FUNCTION_DIR}/node_modules

# Copy in the installer node modules
COPY --from=installer ${FUNCTION_DIR}/node_modules ${FUNCTION_DIR}/node_modules

# Copy the built package
COPY ./dist ./dist/

# If this directory does not exist, lambda shows an annoying warning
RUN mkdir -p /opt/extensions

ENTRYPOINT [ "/usr/local/bin/npx", "aws-lambda-ric" ]
CMD [ "dist/index.handler" ]  

```

**Comment by dl748** [comment]
> @EHadoux please consider to familiarize with tree-shaking and therefore the fact that you cannot do tree-shaking (and minification as well as optimization of the SDK for the specific node.js version of the container) on the provided SDK, thus importing more than needed affecting actually both cold starts and performance.
> 
> With sdk v2 the problem is even bigger than with v3 as it’s modularity is worse.
> 

Not to mention always HORRIBLY out of date.

At the time of this writing, lambda, at least in US-EAST-1, are running version 2.1001.0. The current version according to https://github.com/aws/aws-sdk-js/blob/master/CHANGELOG.md is 2.1069.0

The other reason is that what you install, lint, and run tests against (locally or through build pipeline), are almost NEVER the version that's installed in the lambdas. So unless you include your own SDK, you always run the risk of either A. It not working or B. crashes altogether.

I find including the SDK to be completely a waste for anything production worthy.

I'd really like the lambda system to have global AWS repository of layers that you could include but don't have too.  AWS-SDK would be one of them. I do find it useful to have to run feature tests in lambdas, where I write the code directly in their editor.

**Comment by samchungy** [comment]
The whole purpose of v3 was to allow consumers to only import parts which you want to reduce the bundle size, so including all of the v3 modules would be defeating the purpose of it.

**Comment by jakub-g** [comment]
@jtuliani semi-related but it would be nice if the AWS docs mentioned exact version of node that is running, and not only "12.x" or "14.x"; or document the policy -- how often are those minor versions updated? Are they reasonably up-to-date with the most recent node release in a given line?

Compare [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-nodejs.html) vs [GitHub actions runners specs](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md)

**Comment by tjanczuk** [comment]
@michaelhays Thanks, link fixed!

**Comment by tjanczuk** [comment]
Thank you @jthomerson! I noticed increased latency of function creation but no noticeable difference in startup latency or function execution time. Having said that, it was not a very scientific measurement.

**Comment by 01walid** [comment]
Nice initiative @tjanczu <https://github.com/tjanczuk>!

However, those depending on AWS lambda official node support through other
vendors like Vercel are left with no option, it's still important and
crucial for AWS to officially support this.

Imagine Vercel still being stuck with Node v14...

Disclaimer: I do not work at Vercel, I'm just one of their users.

**Comment by riderx** [comment]
Thanks @tjanczuk for the huge work!
I have asked vercel on Twitter if they can let us use it instead of official one.

It can be a suitable alternative, unlock most of us, and remove pressure of AWS 
https://twitter.com/martindonadieu/status/1493727339288670208?s=20&t=mDwKVHjjdoPTeQ2vQLuPQQ

**Comment by halvardssm** [comment]
I also would prefer to not have any packages installed by default as this will add unnecessary bloat to the runtime. Using modern tooling with tree shaking and bundling should in most cases be the optimal solution. I would however see that an official separate layer could be provided for certain use cases (maybe this could be a part of the build process for v2 and v3 to create these?).

**Comment by ascorbic** [comment]
Thanks @riderx. We're already looking into this, and great work @tjanczuk. Our main concern right now is compatibility, and maintaining confidence that the bootstrap code works consistently across versions. We would likely need thorough testing of each version before we could deploy it.

**Comment by michaelhays** [comment]
Vercel users, go upvote my comment [here](https://github.com/vercel/community/discussions/37#discussioncomment-2200410) to let Vercel know that adding support for [fusebit/everynode](https://github.com/fusebit/everynode) would be very helpful.

**Comment by GeoffreyEmerson** [comment]
> 🎉 🎉 🎉 🎉 https://nodejs.medium.com/node-js-16-available-now-7f5099a97e70
> 
> PLEASE don't wait a year to implement this in a new base image!

Narrator: It was a year before it was implemented in a new base image.

**Comment by rawpixel-vincent** [comment]
> > 🎉 🎉 🎉 🎉 https://nodejs.medium.com/node-js-16-available-now-7f5099a97e70
> > PLEASE don't wait a year to implement this in a new base image!
> 
> Narrator: It was a year before it was implemented in a new base image.

and node 18 LTS is in less than a year, at this pace maybe start planning the base image / runtime for node 18...
but node 16 will be great too!!! following LTS release pace would be perfect.

**Comment by moroine** [comment]
I really don't understand why there is no progress on this, this is a shame from AWS

**Comment by jthomerson** [comment]
@thematan your comment is unnecessary and rude. It has no place here.

While I will publicly say that I think AWS is really dropping the ball here, that's AWS _as an organization_. When you call someone "losers" you're talking about _the people_. Remember that there's a lot of really good people at AWS who are working their hardest. But they have to work within the priorities given to them by their management. What we need to be discussing here is _prioritization_, based on customer needs. This fits with one of AWS' core values - customer-obsession. What we *don't* need to be doing is bad-mouthing the people who are working hard there, just trying to feed their families like all of us. I'd appreciate if you would consider removing your comment.

**Comment by flavianh** [comment]
Hi think @jthomerson said it all. Let's end the string of written complaints here and now if not bringing anything new to the table. And if you do want to complain nevertheless, find the comment you most agree with and upvote it.

**Comment by revmischa** [comment]
> AWS are losers

![stern_cat](https://user-images.githubusercontent.com/245131/155350788-fb7643e6-8a6e-4473-afcd-e4fbac236901.jpg)

**Comment by danthedaniel** [comment]
To AWS developers: I think you're all doing a great job.

To AWS managers: It's time to listen to your developers and customers and prioritize this issue.

**Comment by fitzgerald1337** [comment]
> Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
> 
> We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
> 
> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

this is the only reply that matters. day 24 of 90. 66 more days until the deadline.

may the 4th be with AWS devs.

**Comment by shellscape** [comment]
@fitzgerald1337 where did you see that reply?

**Comment by shellscape** [comment]
> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

@jtuliani I'd love to see the team responsible for these images publicly commit to a release scheduling that follows Node LTS' own schedule. AWS is currently behind GCP, CloudFlare, Vercel, and Supabase (in beta) with regard to LTS Node version support. These are the kinds of gaps in community engagement and trust building that lead independent and smaller shops to competitors, and eventually pull away larger customers. Such a commitment would enshrine reliability and trust for the community around an extremely popular platform.

**Comment by talaikis** [comment]
Already 6+ months after LTS :-D

**Comment by rawpixel-vincent** [comment]
+1 

following

👍🏻

**Comment by wc-davide** [comment]
> PLEASE don't wait a year to implement this in a new base image!

_getting closer 😔_

**Comment by jcollum** [comment]
Node 16 transitioned to LTS 5 months ago and AWS still doesn't support it? Surprising and kinda lame really. At least put up a blog post discussing why this is so late. It's not like new Node versions are a surprise. They're released on a very predictable cadence. 

No bueno AWS.

**Comment by rawpixel-vincent** [comment]
following +1

**Comment by rawpixel-vincent** [comment]
<img width="614" alt="Screen Shot 2565-03-23 at 02 00 48" src="https://user-images.githubusercontent.com/22284209/159556007-41751c6a-3adb-4100-a7c1-fa4232c16025.png">

**Comment by rawpixel-vincent** [comment]
locked down much
https://github.com/opensearch-project/OpenSearch/issues/1400#issuecomment-948921005

**Comment by delta-9** [comment]
for the stockholm haters just making it clear that I mistakenly posted this from my pro account, so hey 👋 I'm vincent, and not worth to elaborate when the reason for the trolling are obvious

**Comment by digitaltoast** [comment]
I have literally no idea what you are drivelling on about, but I'm guessing you're going to wake up with a big hangover and a lot of embarrassment in the morning...
So this isn't wasted post Connor I just want to add my thanks to @tjanczuk for doing what the world's largest tech company couldn't.

**Comment by jcollum** [comment]
> following +1

There's a "subscribe" button @rawpixel-vincent -- been there for ohhhh 5 years now?

**Comment by tjanczuk** [comment]
@digitaltoast I appreciate it, thank you. Any time. Every node ;) https://github.com/fusebit/everynode

**Comment by jcollum** [comment]
> @digitaltoast I appreciate it, thank you. Any time. 

huh? 

> Every node ;) https://github.com/fusebit/everynode

Hmm that actually might be useful...

**Comment by Prophet32j** [comment]
following. I've been waiting for a v16 runtime for a while now.

**Comment by hayksaryan** [comment]
@Prophet32j and everyone who wants to follow a thread/discussion on Github. Use the button in the screenshot to do so instead of commenting. 
![image](https://user-images.githubusercontent.com/52959001/159931331-d0a33faa-7108-46a6-ada9-6b2be92e480b.png)

**Comment by bartclarkson** [comment]
Thank you @tjanczuk .

For others using SAM Templates for local development, here is a fairly streamlined example of using the latest Node Version from fusebit.  Note that the Runtime is rendered inconsequential by the Layer list.

It takes a second the first time for the layer to be pulled down and the docker image built.  Then you're good.  

```
AWSTemplateFormatVersion: 2010-09-09
Transform: 'AWS::Serverless-2016-10-31'
Description: Fusebit for the win

Globals:
  Function:
    Runtime: nodejs14.x
    MemorySize: 128
    Timeout: 15

Resources:
  ServerlessHttpApi:
    Type: AWS::Serverless::HttpApi
    Properties:
      Description: Amazon API Gateway HTTP API + Lambda Function.
      ProtocolType: HTTP
      CorsConfiguration:
        AllowOrigins:
          - '*'
        AllowMethods:
          - GET

  DefaultLambdaFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: handlers/testing.handler
      Layers:
        - "arn:aws:lambda:us-east-1:072686360478:layer:node-17_8_0:1"
      Events:
        Other:
          Type: HttpApi
          Properties:
            Path: /{proxy+}
            Method: get
            ApiId: !Ref ServerlessHttpApi
```

**Comment by jlarmstrongiv** [comment]
@jtuliani

> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

Your https://github.com/aws/aws-lambda-nodejs-runtime-interface-client needs to be upgraded with the Node 16 release.  I strongly encourage you to keep it up to date https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/44#issuecomment-1032928940 and merge https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/pull/8 to avoid complex and long build steps.  In addition, I would like to see compatibility stated for AWS Linux 2022.

It is vital to maintain parity between aws lambda zip and aws lambda docker deploys.

**Comment by mooyoul** [comment]
@tjanczuk Don't do Promise instance check using `result instanceof Promise`. Global Promise can be mutated by third party libraries like Bluebird and Instrumentation support libraries like async-listener.  Official Lambda runtime had issue with their first Node.js 8 lambda runtime release. See https://github.com/aws/aws-xray-sdk-node/issues/27#issuecomment-380081229 for historical records.

**Comment by mooyoul** [comment]
@jtuliani 

> That's an open question we're working through right now. If you have strong opinions, we'd be interested to hear them! For example, is it useful to ship the SDKv2 and SDKv3 side-by-side, so you can upgrade functions to Node 16 without switching SDK? Or is it OK to ship the SDKv3 only, and require an upgrade to Node 16 to also mean upgrading to SDKv3? (Obviously this only applies to functions using the bundled SDK.)

IMHO It would be great if both SDK are available in newer Lambda runtime (not meant for Container Image)

Keeping SDK V2 will give users backwards compatibility - Most programs will run in newer runtime without additional migration. Previously Node.js had multiple breaking changes (e.g. Stream implementation changes in v0.8 ~ v0.12), but nowadays most core modules are stable or frozen so same deployment artifact will be okay in newer Node.js runtime.
but if Node.js 16 Lambda runtime removes AWS SDK V2, it forces users to change their applications - It forces developers to include AWS SDK packages to their artifact. and No one likes **unexpected** change. I believed that every AWS Lambda runtime has "bundled" packages. but now i realize it can be broken - Also I remember when AWS Lambda changed base OS to Amazon Linux 2. A lot of built-in system packages were removed in Amazon Linux 2 and Migrating existing softwares were super annoying - I had to rebuild build environment, rebuild native dependencies (e.g. ImageMagicK, OpenSSL dependent softwares), rebuild native addons, and repackage everything - It was totally nightmare. 

Introducing AWS SDK v3 should be worthless because AWS SDK v3 has more modular and bundler friendly that AWS SDK v2. but It would be worth if function package is heavy (e.g. deployment size exceeds 250MB with AWS SDK V3).

For performance-wise issues, I don't think that would make negative performance impact. because I think most "AWS-provided" official Lambda runtime would be cached in their compute infrastructure. Also Node.js runtime does not load dependences unless you importing it.


and for Runtime Container Image - Just give choices to developers.

**Comment by tjanczuk** [comment]
@mooyoul Thanks for the pointer, we will fix this.

**Comment by dl748** [comment]
I would just turn all results into a promise. I've done this in several libraries.

```
try{
  tmp = await Promise.resolve(handler(body, context, done))
} catch(e) {
  reject(e)
}
```
That way I don't have to do special processing for a function that may or may not return a promise, it just naturally "unwraps".

**Comment by halvardssm** [comment]
@jtuliani 

> That's an open question we're working through right now. If you have strong opinions, we'd be interested to hear them! For example, is it useful to ship the SDKv2 and SDKv3 side-by-side, so you can upgrade functions to Node 16 without switching SDK? Or is it OK to ship the SDKv3 only, and require an upgrade to Node 16 to also mean upgrading to SDKv3? (Obviously this only applies to functions using the bundled SDK.)

I would prefer if the Node Lambda image is as small as possible, that means it should not include SDK V2 (or V3 for that matter). I understand the argument "it forces users to change their applications", but for a Node-Lambda version change, that is to be expected, and also should be the case. I rather remove `aws-sdk` from my exclude list in the bundle script or add it to the dependencies list, than have it included in the runtime. 

As a solution for those who do not want to change their build output, AWS could potentially provide an official and public layer containing the SDK v2, and maybe also provide a layer for each V3 package.

The above is my preferred outcome, but if you decide to include the SDK in the runtime rather than providing layers, I would prefer that both versions would be included.

**Comment by AaronSterlingGENEICD** [comment]
I'd prefer a combination of those two approaches:

1. Lambda by default builds with no layers, as light as possible.'
2. The console (and CLI) allows the addition of the most commonly added layers with a couple mouse clicks. Example: a checkbox to add AWS SDKv2 as a layer.
3. Stretch goal: a select dropdown for SDKv3, that allows you to add layers for the clients you need in a fast but fine-grained way.

**Comment by dl748** [comment]
@halvardssm
> I would prefer if the Node Lambda image is as small as possible, that means it should not include SDK V2 (or V3 for that matter). I understand the argument "it forces users to change their applications", but for a Node-Lambda version change, that is to be expected, and also should be the case. I rather remove `aws-sdk` from my exclude list in the bundle script or add it to the dependencies list, than have it included in the runtime.

Why would you EVER put it in an exclude list. That's a recipe for problems. If you are building against the current runtime version (2.1001.0) you are missing out on 112 releases (2.1113.0) including bug fixes. If you are building against the current, you are testing your code against a newer version but running it on a way older version. Unless its TEST code, you should always use the sdk for which it has been tested against. I would trim other code than use a library you are not running tests against.

**Comment by shellscape** [comment]
Is there another thread or discussion you all can jump into to hash this out? I'm subscribed here for updates from AWS on the original post.

**Comment by dl748** [comment]
@shellscape

It does pertain to the original post, as it was asked what did we want in the image of Node 16. I, along with several people, hope they remove the javascript SDK from the node image. It may not fit your limited scope of what pertains to the post, but it does to many other people.

**Comment by shellscape** [comment]
Cool, cool. Well then.

**Comment by halvardssm** [comment]
> If you are building against the current runtime version (2.1001.0) you are missing out on 112 releases (2.1113.0) including bug fixes

Exactly why I would like to not have it included in the runtime, and rather have layers that can be versioned and updated alongside the SDK.

@dl748 funnily enough, this is sometimes the only solution if your function is too large as v2 does not allow for tree shaking.

**Comment by chenrui333** [comment]
Hello, why did this get closed? I am subscribing in here for the runtime image version update, but I did not see that actually happened, right?

**Comment by paul-uz** [comment]
> Cool, cool. Well then. 

This issue is not resolved. Why have you closed it?

**Comment by chenrui333** [comment]
I dont want to lose the context in this issue, @shellscape do you mind re-opening it? Thanks!

**Comment by shellscape** [comment]
Folks, AWS isn't interested in resolving this in any semblance of a timely manner. We're 8 days shy of a year since I created the issue. Too many people who are commenting on this issue don't understand the concept of "off topic," and I've grown tired of notifications from replies by people like @dl748, who's reply you can thank for prompting me to close this. I'm not going to reopen this, and I won't see further replies. You're welcome to open your own issues for tracking this.

**Comment by achuinard** [comment]
@shellscape couldn't you just unsubscribe from more updates?

**Comment by Negan1911** [comment]
No, no, he had to throw his giant rant onto us

**Comment by kf6kjg** [comment]
I get it: I see that the OP was told that he didn't understand what the topic was. While it may have grown beyond the original scope, the OP gets to decide what the scope is.

While closing this is petty, it's within the OP's rights. Especially when this request is getting so long in the tooth.

Maybe one of us should open a new post and link it back here? Just be warned: you'll get a lot of comments, be ready for them! :D - haha Just as I was writing this a few people did!

**Comment by jlarmstrongiv** [comment]
@shellscape https://github.com/aws/aws-lambda-base-images/issues/14#issuecomment-1029958936

> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

**Comment by fitzgerald1337** [comment]
yeah, i totally get why OP would want to close this shit show of a thread; i totally get why the people following it want it to stay open; i think ultimately what it comes down to is that AWS devs/reps (@jtuliani) should have an official thread and shouldn't assume that OP's would automatically serve as an official one.

please, AWS, get a better dev advocate(s) on this.

**Comment by forty** [comment]
What I suggest is that everyone who would need NodeJS 16 support on Lambda to regularly reach out to AWS support to remind them we need this feature. I think this is a best way than going through this GH issue.

Personally I don't really mind what is included in the image, I just want an AWS supported way of running NodeJS 16 in Lambda, and I want it ASAP, because we migrated our other systems to NodeJS 16, and we are not fond of juggling with several versions of Node.

**Comment by chenrui333** [comment]
> What I suggest is that everyone who would need NodeJS 16 support on Lambda to regularly reach out to AWS support to remind them we need this feature. I think this is a best way than going through this GH issue.

Agree, I think everyone should reach out to the AWS TAM and reference this issue as source. So that we have both parts running (internal TAM and public nudge as well as transparency).

**Comment by ennioVisco** [comment]
> > > Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
> > > We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
> > > If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.
> > 
> > 
> > this is the only reply that matters. day 24 of 90. 66 more days until the deadline.
> > may the 4th be with AWS devs.
> 
> Day 50 of 90. 40 to go. This will likely end to be more than a year from the release.

Day 70 of 90. 20 to go (?)

**Comment by ES-Six** [comment]
> Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
> 
> We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
> 
> If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.

I have a suggestion :
- It can help a lot of people to plan a migration to the next NodeJS LTS in AWS Lambda if AWS could provide a kind of "implementation roadmap" to announce a date (or at least a month or a quarter) when the next LTS version will be available in AWS Lambda, exactly like NodeJS roadmap does for nodeJS releases, but no need to be so precise (a month or at least a quarter may be sufficient). So customers will be able to know that AWS Team is aware of the nodeJS LTS upgrade and is working on it.

**Comment by tapz** [comment]
This is the reply I got from AWS technical support:

> The team has shared that, Node 16 runtime support in Lambda is being worked upon, and they have plans to release it in MAY-2022. You can keep an eye on the following links for more information and updates on the same :
> ...
> Additionally, the team is aware of the feedback we have been receiving from customers, that they would like to see new runtimes released more promptly. Please be assured that the service team will take the feedback very seriously and work towards improving the Lambda service for better customer experience. Hence, the team has shared that they are reviewing the runtime release processes and resourcing, so we can better meet these expectations in the future.

**Comment by ES-Six** [comment]
> This is the reply I got from the AWS technical support:
> 
> > The team has shared that, Node 16 runtime support in Lambda is being worked upon, and they have plans to release it in MAY-2022. You can keep an eye on the following links for more information and updates on the same :
> > ...
> > Additionally, the team is aware of the feedback we have been receiving from customers, that they would like to see new runtimes released more promptly. Please be assured that the service team will take the feedback very seriously and work towards improving the Lambda service for better customer experience. Hence, the team has shared that they are reviewing the runtime release processes and resourcing, so we can better meet these expectations in the future.

Vary good, now we have a more precise release period, now we now nodeJS v16 is planned to be release in AWS Lambda in May 2022. Thanks for sharing this information.

**Comment by chenrui333** [comment]
Log an issue for node v18, #47

**Comment by davecarlson** [comment]
> 🎉 🎉 🎉 🎉 https://nodejs.medium.com/node-js-16-available-now-7f5099a97e70
> 
> PLEASE don't wait a year to implement this in a new base image!

They waited a year, and more !

**Comment by klhex** [comment]
> Is there a way to create a custom runtime to use the latest version of node? If so, can anyone point me in the right direction?

It seems so: https://github.com/fusebit/everynode

**Comment by tonymporter** [comment]
> 🎉 🎉 🎉 🎉 https://nodejs.medium.com/node-js-16-available-now-7f5099a97e70
> 
> PLEASE don't wait a year to implement this in a new base image!

lol they did!!

**Comment by paul-uz** [comment]
@klhex any idea how to use everynode in Cloudformation, specifically with a AWS::Serverless::Function resource?

**Comment by paul-uz** [comment]
@klhex I thought the same thing, and tried implementing it, but I get the following error when trying to run the Lambda

> Unknown application error occurred
Runtime.ErrorLoadingHandler

No other information given to help me debug. I made sure my other Layer (node modules) had `provided` as a runtime compatability, and tried various things with the Handler setting, but nothing seems to work.

**Comment by klhex** [comment]
> @klhex I thought the same thing, and tried implementing it, but I get the following error when trying to run the Lambda
> 
> > Unknown application error occurred
> > Runtime.ErrorLoadingHandler
> 
> No other information given to help me debug. I made sure my other Layer (node modules) had `provided` as a runtime compatability, and tried various things with the Handler setting, but nothing seems to work.

I see, that's a pity. (I haven't tried it out at all yet, since we are only allowed to use AWS runtimes and therefore have to wait until AWS releases a new nodejs runtimes...)

But I saw that you raised a [ticket](https://github.com/fusebit/everynode/issues/11) for this issue. I hope that someone will be able to help you out.

**Comment by paul-uz** [comment]
@klhex sadly im not hopeful, as I dropped in to their Discord to ask and it's dead.

Fingers crossed for the offical AWS runtime.

**Comment by BeeeQueue** [comment]
> @klhex any idea how to use everynode in Cloudformation, specifically with a AWS::Serverless::Function resource?

I imagine it wouldn't be much different from how you do it in terraform:

```tf
resource "aws_lambda_function" "test" {
  #...

  runtime     = "provided"
  layers      = ["arn:aws:lambda:eu-north-1:072686360478:layer:node-17_0_0:1"]

  environment {
    variables = {
      AWS_NODEJS_CONNECTION_REUSE_ENABLED = "1"
    }
  }
}
```

**Comment by paul-uz** [comment]
@BeeeQueue I tried adding it to my Cloudformation configs, and got nothing but unhelpful errors (see above)

**Comment by jtuliani** [comment]
Thank you everyone for the continued engagement and feedback on this issue. The AWS Lambda Node 16 runtime implementation is in progress, and we are currently on track to launch in mid-May. Our Node 16 runtime will continue to use AWS SDKv2, and we plan to transition to AWS SDKv3 with Node 18.

**Comment by paul-uz** [comment]
@jtuliani when are you looking to release Node 18 support? Node 16 is already yesterdays news.

**Comment by fromtheexchange** [comment]
@jtuliani thank you for the update!  Will you also be updating https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/44#issuecomment-1032928940 ?  I open-sourced [prebuilt packages](https://github.com/fromtheexchange/aws-lambda-ric) for amazonlinux:2022, but I need parity between your default aws-lambda-ric and the docker aws-lambda-ric, especially for ESM support.

**Comment by o5** [comment]
> The AWS Lambda Node 16 runtime implementation is in progress, and we are currently on track to launch in mid-May

@jtuliani Could you specify the year, please?

**Comment by fitzgerald1337** [comment]
> > Thank you everyone for your feedback regarding AWS Lambda support for Node 16. Work is in progress, and we are planning to ship a Node 16 managed runtime and base container image within the next 90 days.
> > 
> > We also apologize for the confusion caused by locking this thread. We are addressing the gap in our process that caused it to happen. The feedback and engagement from this community is very valuable to us, and we will keep iterating on improving how we engage here.
> > 
> > If you have any suggestions for specific features or enhancements you would like to see included in our Node 16 support, please let us know on this thread.
> 
> this is the only reply that matters. day 24 of 90. 66 more days until the deadline.
> 
> may the 4th be with AWS devs.

aaaaaaaaaaaand deadline missed

**Comment by jtuliani** [comment]
Update: There are multiple parts to a runtime launch. There's the managed runtime itself, the container base image, plus support in the console, SAM CLI, AWS CLI, CloudFormation, various AWS SDKs, etc. In previous releases we have coordinated all of these to land on launch day. For the Node 16 release, we're experimenting with a different approach, with the various components landing separately. A release candidate of the 'nodejs16.x' runtime identifier is already deployed (you can try it today if you deploy via the Lambda APIs directly), with the other components in progress. Once everything is complete, we'll make the GA announcement.

@fromtheexchange That's a great question regarding the alignment of RICs. This is something we currently plan to address in Node 18.

@The-Code-Monkey Let's discuss Node 18 in #47, so we keep that information in one place.

**Comment by The-Code-Monkey** [comment]
@jtuliani thanks for the update, I will have a play with it

**Comment by ascorbic** [comment]
Confirmed that this works on Netlify functions. Set `AWS_LAMBDA_JS_RUNTIME` to `nodejs16.x` and it works: https://upbeat-snyder-05e8fe.netlify.app/api

**Comment by JuanM04** [comment]
It also works on Vercel with the new Build Output API

![image](https://user-images.githubusercontent.com/16712703/167412489-813a80fd-6ddc-4d3a-968f-c07181f527e7.png)

**Comment by jtuliani** [comment]
Thanks @forty for flagging that error message, I'll get that looked at.

Regarding the exact ETA: there are a lot of moving pieces to a runtime release (console, docs, SDK, SAM, CDK, CLI, CFN, etc). There are also multiple teams involved. Plans for any one of these teams can change at short notice, for example due to a live site issue, a deployment roll-back, or surprises such as Log4J2. So, I'm sure you'll understand why I'm holding off from providing an exact schedule. However, we are getting closer :)

**Comment by pgrzesik** [comment]
Thanks for a swift response @jtuliani - that's totally understandable 👍 Fingers crossed for a smooth release process 🤞 Serverless Framework users are also very eager to try the new runtime - https://github.com/serverless/serverless/pull/11049 🙌

**Comment by jtuliani** [comment]
Node 16 support in Lambda is now GA: https://aws.amazon.com/about-aws/whats-new/2022/05/aws-lambda-adds-support-node-js-16/

**Comment by joebowbeer** [comment]
@jtuliani wrote:
> Node 16 support in Lambda is now GA: https://aws.amazon.com/about-aws/whats-new/2022/05/aws-lambda-adds-support-node-js-16/

In this announcement, the following sentence caught my attention:

> Node.js 16 is the latest long-term support (LTS) release of Node.js and will be supported for security and bug fixes until April 2024.

Given that Node 16 reaches EOL in [September, 2023](https://nodejs.org/en/about/releases/), how is AWS planning to make security and bug fixes after that date?

**Comment by davecarlson** [comment]
What are the time scales for this ?

**Comment by jtuliani** [comment]
See #47 for Node 18 support in Lambda
